meilisearch-core/src/facets.rs

```diff
@@ -0,0 +1,355 @@
+use std::hash::Hash;
+use std::ops::Deref;
+
+use cow_utils::CowUtils;
+use either::Either;
+use heed::types::{Str, OwnedType};
+use indexmap::IndexMap;
+use serde_json::Value;
+
+use meilisearch_schema::{FieldId, Schema};
+use meilisearch_types::DocumentId;
+
+use crate::database::MainT;
+use crate::error::{FacetError, Error};
+use crate::store::BEU16;
+
+/// Data structure used to represent a boolean expression in the form of nested arrays.
+/// Values in the outer array are and-ed together, values in the inner arrays are or-ed together.
+#[derive(Debug, PartialEq)]
+pub struct FacetFilter(Vec<Either<Vec<FacetKey>, FacetKey>>);
+
+impl Deref for FacetFilter {
+    type Target = Vec<Either<Vec<FacetKey>, FacetKey>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl FacetFilter {
+    pub fn from_str(
+        s: &str,
+        schema: &Schema,
+        attributes_for_faceting: &[FieldId],
+    ) -> Result<Self, FacetError> {
+
+        let parsed = serde_json::from_str::<Value>(s).map_err(|e| FacetError::ParsingError(e.to_string()))?;
+        let mut filter = Vec::new();
+        match parsed {
+            Value::Array(and_exprs) => {
+                if and_exprs.is_empty() {
+                    return Err(FacetError::EmptyArray);
+                }
+                for expr in and_exprs {
+                    match expr {
+                        Value::String(s) => {
+                            let key = FacetKey::from_str( &s, schema, attributes_for_faceting)?;
+                            filter.push(Either::Right(key));
+                        }
+                        Value::Array(or_exprs) => {
+                            if or_exprs.is_empty() {
+                                return Err(FacetError::EmptyArray);
+                            }
+                            let mut inner = Vec::new();
+                            for expr in or_exprs {
+                                match expr {
+                                    Value::String(s) => {
+                                        let key = FacetKey::from_str( &s, schema, attributes_for_faceting)?;
+                                        inner.push(key);
+                                    }
+                                    bad_value => {
+                                        return Err(FacetError::UnexpectedToken { expected: &["String"], found: bad_value.to_string() })
+                                    }
+                                }
+                            }
+                            filter.push(Either::Left(inner));
+                        }
+                        bad_value => return Err(FacetError::UnexpectedToken { expected: &["String", "Array"], found: bad_value.to_string() }),
+                    }
+                }
+                return Ok(Self(filter));
+            }
+            bad_value => Err(FacetError::UnexpectedToken { expected: &["Array"], found: bad_value.to_string() }),
+        }
```

Comment:
Could you create a contructor for the unexpected token, this will reduce the amount of code here.